services/src/main/java/io/grpc/protobuf/services/ProtoReflectionService.java

```diff
@@ -140,6 +150,21 @@ private ServerReflectionIndex updateIndexIfNecessary() {
     return requestObserver;
   }
 
+  private static final class ServerCaptureInterceptor implements ServerInterceptor {
+    private final AtomicReference<Server> captor;
+
+    ServerCaptureInterceptor(AtomicReference<Server> captor) {
+      this.captor = captor;
+    }
+
+    @Override
+    public <ReqT, RespT> Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
+        Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+      captor.set(call.getServer());
```

Comment:
I don't think it would be obvious that you must use a different reflection service instance for each server. Can we instead add the Server to the Context and handle multiple servers? (WeakHashMap may make sense)

Reply:
As our offline discussion, added a WeakHashMap to cache indexes for servers. This supports a single reflection service instance to be shared across multiple servers, without recomputing its service index upon each RPC.

I don't think adding the cache to the Context is necessary. Did I miss anything?

Reply:
No, this AtomicReference doesn't work. This is not thread-safe. If two reflection RPCs come in at the same time, the responses may be for the wrong Server. Add the `Server` to Context here (not the cache; although adding the cache entry instead would be "fine" functionally).